docs/channel.md

```diff
@@ -424,6 +424,47 @@ Y
 
 See also: [fromList](#fromlist) factory method.
 
+(channel-topic)=
+
+### topic
+
+:::{versionadded} 23.11.0-edge
+:::
+
+:::{note}
+This feature requires the `nextflow.preview.topic` feature flag to be enabled.
+:::
+
+The `topic` method is used to create a "topic" channel, which is a queue channel that can receive items from multiple sources.
```

Comment:
Maybe you want to describe "topic" channels as a new channel type alongside queue and value channels. Since a topic channel seems to behave like a queue channel, I described it in this way

Reply:
Good point. Surely it deserves to be expanded. Another possibility could be introduce the "broadcast" channel type, along "queue" and "value". A broadcast channel can even many writers and many readers (opposed to a queue channel than can have exactly one write and reader) and it's identified by a "topic" name.

Reply:
I like "broadcast" more than "topic", sounds to me like more appropriate jargon in the scope of Nextflow

Reply:
Broadcast is not specific enough IMO because a value channel is also a broadcast (it can have many readers). Even the underlying GPars class is `DataflowBroadcast`. The topic channel is distinct because it can have many *writers*, but I don't know of any special term for that. I will see if I can find something from stream processing or digital circuits terminology...

Reply:
I didn't find anything beyond the [event bus](https://dzone.com/articles/design-patterns-event-bus) pattern. On further reflection, I don't think we need to distinguish topic channels as a special channel type. A topic channel is just shorthand for a mix operation:

```groovy
ch_foo = mix(foo1, foo2, foo3, foo4, foo5)
```

It's just a bunch of queue channels, and the "topic" is what brings them together.

I'm open to other words than topic if we can find a better one. Some alternatives include "category", "label", "tag"... but label and tag are already concepts in Nextflow and category is too broad IMO. "metadata" implies that the topic channel can only be used to collect metadata, but that need not be the case. I like the idea of describing the topic channel as an event bus, but any channel could be called a "bus".

How about... [`Channel.mixer()`](https://en.wikipedia.org/wiki/DJ_mixer)

Reply:
I think you were right it should be introduced a new channel definition because in several places it's said that a "queue" channel only allows one writer and one reader, therefore this is an important change in the nextflow dataflow model. 

I'd keep aligned with gpars definition: 
* DataflowQueue -> queue channel 
* DataflowValue -> value channel 
* DataflowBroadcast -> broadcast channel 

And a "broadcast" channel can be identified by "topic" name. The topic is common in event based systems, and [Kafka made](https://dzone.com/refcardz/apache-kafka-patterns-and-anti-patterns) it popular

Reply:
I think we are misunderstanding each other. Broadcast is *not* a good term to use because it only means multiple readers, not necessarily multiple writers. I'm talking about the meaning of the word, not the semantics of `DataflowBroadcast`.

Also, Nextflow uses `DataflowBroadcast` under the hood to implicitly fork a queue channel when it is read multiple times in a pipeline, so there isn't really a one-to-one mapping between GPars and Nextflow channel types. In reality, both queue and value channels can have multiple readers in Nextflow, so they can all be "broadcast".

There isn't really a term for a channel that can have multiple sources, I guess because there are different terms for different kinds of aggregation (mix, collect, gather, concatenate, sum, etc). Since we are simply "mixing" the source channels (as in the `mix` operator), the best term I can think of is a "mixer" channel. So I recommend we declare a new "mixer" type or just don't call it a new type at all.

Reply:
I see better your point, however, I still claim "broadcast" fits better here. It's true that's not perfectly aligned with the real semantics but at least it's a term more common in the message-oriented systems. 

Also if you notice gpars docs, they use DataflowQueue both for one-to-many and many-to-many 

http://gpars.org/1.2.1/javadoc/index.html?groovyx/gpars/dataflow/DataflowBroadcast.html

Reply:
> Broadcast tells you nothing about the number of writers

I know, but a simple word cannot capture the semantics. The important thing is that it's evocative enough to give an idea of the different behaviour. The docs will give a formal definition of the semantics and how it can be used

Reply:
Paolo, that does not make any sense 😆 

Besides, we can do better than merely "evocative" -- we can actually use a correct word! That word is "mixer"

Reply:
To me "mixer" make no sense 🤷‍♂️, and the consistency with gpars terminology it's an important +1 for me.

Reply:
Under the hood is a composition of `mix` operator, however the logic model is distinct

Reply:
I think it's inconsistent of you to treat the alignment with DataflowBroadcast as such a benefit while brushing aside the mix operator as merely an implementation detail. But I will put that aside because I don't like "mixer" for other reasons. My actual preference is to not declare a new channel type at all.

I think I would rather just call it a "topic" channel. The multiple writers and "broadcast" are not the important aspect -- any channel can be broadcast in Nextflow, and operators can have "multiple writers" -- but the implicit aggregation of inputs based on a _topic_. And as you have said, "topic" is a well-established term from the domain of stream processing.

Look over the docs I have written. I have described what a topic channel is and how it is used in a pipeline. Do you really think these docs will be improved by adding another term aside from "topic", especially one that is _not correct_ ?

Reply:
Just read the thread. 

I am not super convinced of "topic", but I cannot think of a better one neither. At first I was liking "broadcast", but Ben's comments around inconsistencies that can turn confusing to users/devs is fair. 

However, I do think it should be advertised as a new channel type, at the same level of value and queue. The distinctive trait is indeed the ability to collect input from multiple processes. (this is opposite to considering it a specialisation of the queue type)

So, I would go for the `topic` channel, which is a new third type of channel.

Reply:
To me, the key distinguishing feature of the topic channel is not the multiple sources, but the ability to collect those sources *implicitly* via a matching name which is the "topic".

I don't see it as a new channel type, but an operation that produces a queue channel. It is essentially an implicit "mix" operation, but we don't create a new channel type for every operator like mix, collect, concat, etc, we just call them channel operators. Or consider the `Channel.watchPath()` factory, which emits file events for a path or glob -- it is doing a special kind of operation, but ultimately produces a queue channel. So I think the `topic` factory should be understood as an operation on queue channels.

Maybe we should call it `Channel.fromTopic` to be consistent with `fromPath`, `fromList`, etc

Reply:
Thanks for this perspective Ben, I get your point, it makes sense to me now: in the end the multiple writers end up pouring their output in waht is a single queue - the description you made of implicit multiple mixes got me.

@pditommaso ?

Reply:
I love `fromTopic`, will comment further down

Reply:
Quoting Paolo's comment which got mixed up:

> I'm fine to stay with "topic" channel definition, but they should be documented as a new channel type because they have different semantics (many writers).
>
> Thus it should be `channel.topic`, not `fromTopic`

Thinking more on this, I still think we should call it `Channel.fromTopic` and describe it as an operation that creates a queue channel rather than a new channel type. Marco, Phil, and I seem to be in agreement on this, but since this thread has meandered and the meeting didn't reach a solid conclusion, here is my argument put concisely:

A channel topic is literally an operation on queue channels. There is no new channel type under the hood, just queue channels coming in and a queue channel going out. It uses the DataflowBroadcast only to support multiple readers (in the same way as queue channels) and it uses the *mix operator* to support the multiple writers. In fact many operators support multiple writers, so that alone is not enough to warrant a new channel type. The implicit linking via topic is more unique, but when I tried to document it as a new channel type, I just found it unnecessary and more confusing.

Reply:
> it uses the mix operator to support the multiple writers. In fact many operators support multiple writers, so that alone is not enough to warrant a new channel type.

👍

Reply:
I believe you need to see things from the proper perspective. The most important thing is that the topic channel introduces a different way to compose and think Nextflow channels. 

Instead of having one-to-one, producer-to-consumer messaging, the topic allows many producers to send messages over the same topic to many consumers. 

It doesn't matter it could have been implemented using a composition of `mix` operators or how it's implemented under the door. The topic type is important to highlight the different paradigm that is introduced by this feature. 

I've made a few changes in the docs to reflect this view. In any case, this is marked as experimental, we can always review and changes along the way in future releases.

Reply:
> the topic channel introduces a different way to compose and think Nextflow channels.
> The topic type is important to highlight the different paradigm that is introduced by this feature.

I had this chat with Paolo as well. I now better get the point around proposing a new paradigm, that unlocks new ways for devs to describe their pipelines. 

Overall it was a good discussion, and most importantly it is good to have it as `experimental`, to leave room for upgrades in case we identify the need for them.

Reply:
I do see the other perspective Paolo, after all I started this whole thread, but my thoughts evolved and I no longer think it's necessary to describe channel topics as a new type. Although it doesn't always happen this way, in this case I think the implementation details are quite instructive in how to describe it. If someone figures out to use a channel topic in some way other than an implicit mix, I might be convinced otherwise.

But I'm glad you went ahead and merged it. Better to get the feature out there for users to play with it. We can refine the docs as needed.